compiler/passes/analyses/track_volume.py

```diff
@@ -0,0 +1,252 @@
+import networkx as nx
+
+from compiler.data_structures import Program
+from compiler.passes.analyses.bs_analysis import BSAnalysis
+from compiler.data_structures.ir import *
+
+
+class VolumeTracker(BSAnalysis):
+
+    variable_volume = dict();                # The dict that will store the volumes of relevent variables
+
+    violation_found = False
+
+    _program = None;
+
+    def __init__(self):
+        super().__init__("Volume Tracking") # Sets the name to Volume Tracking?
+
+    def analyze(self, program: Program) -> dict: # The main function of the class
+        print("Now tracking volume...")          # Debugging statement   
+
+        self._program = program
+        
+
+        for root in program.functions:
+
+            print("Functions from graph: ")
+            for node, data in program.bb_graph.nodes(data=True):
+                #print(program.functions[root]['blocks'][node])
+
+                for i in program.functions[root]['blocks'][node].instructions:
+                    #print("type of i: " + str(type(i)))
+
+                    if (type(i) == Phi):
+                        print("Found a Phi node!")
+                        print(i)
+                    #print("i: " + str(i))
+                    #print("defs: " + str(i.defs))
+                    #print("uses: " + str(i.uses))
+
+                    handle(self, i)
+
+                    if (self.violation_found):
+                        print("Exiting...")
+                        break;
+
+                    print(self.variable_volume)
+
+            if (self.violation_found):
+                        print("Exiting...")
+                        break;
+
+        return {'name': self.name, 'result': None} # Not sure what is going on here, but its needed to prevent a crash
+
+
+
+def handle(self, instruction : IR): # The meat of the logic. Just ferries out the functions based on the type of instruction.
+    if (type(instruction) == Dispense):

```

Comment:
IRInstruction has an `op` field that tells you what it is.  You don't need to do `type() == ...`, you can do `instruction.op == IR.[op]`

Reply:
Using Type() is actually very fast and functions well considering all Python variables are self-aware of their typing.